---
layout: default
title: Resume
permalink: /resume/
---
<div align="center">
  <h2>Zachary L. Skidmore</h2>
</div>
<div align="center">
 <a href="https://www.linkedin.com/in/zachary-skidmore-0799653a/"><i class="fa-brands fa-linkedin"></i> LinkedIn </a> | <a href="https://orcid.org/0000-0003-1928-7139"><i class="fa-brands fa-orcid"></i> ORCiD </a> | <a href="https://github.com/zlskidmore"><i class="fa-brands fa-square-github"></i> GitHub </a>
</div>
<hr>

### Professional Summary:
Senior bioinformatics and quantitative scientist with 12+ years spanning Tier 1 academic research and industry. Proven expertise in study design, cancer genomics analysis, and pipeline/software development. Author of high-impact publications and patent co-inventor. Experienced mentor and instructor, developed and taught multiple bioinformatics courses. Seeking a senior or lead bioinformatics scientist position to apply deep technical expertise and expand leadership scope.

### Education:
<div class="container">
  <div class="d-flex justify-content-between mx-auto" style="width: 100%;">
    <div class="text-start">
      <p>University of Illinois<br>Master Of Engineering in Bioinformatics</p>
    </div>
    <div class="text-end">
      <p>Chicago, IL<br>May 2013</p>
    </div>
  </div>
</div>

### Work Experience:
<div class="container">
  <div class="d-flex justify-content-between mx-auto" style="width: 100%;">
    <div class="text-start">
      <p><strong>Delfi Diagnostics</strong><br>Position: Senior Quantitative Scientist, Translational Science, Data Science - Research</p>
    </div>
    <div class="text-end">
      <p>Remote<br>Aug 2022 - Present</p>
    </div>
  </div>
</div>

###### 2023 - 2025 Highlights
- Co-developed and submitted intellectual property for differentiating histological subtypes of lung cancer using cfDNA features [WO2024173277A2](https://patents.google.com/patent/WO2024173277A2/en?oq=WO2024173277A2)
- Lead author and bioinformatic scientist for development and application of ctDNA monitoring on a prospective phase III clinical trial (CAIRO5) using fragmentomics PMID: [39433569](https://pubmed.ncbi.nlm.nih.gov/39433569/)
- Trained and maintained a random forest model to predict ctDNA fraction from plasma
- Engaged with drug development companies in the application and analysis of ctDNA monitoring across multiple cancer types
- Contributed to more than 15+ translational science projects from inception through publication, resulting in numerous domestic and international conference presentations/abstracts and manuscripts (under review)
- Contributed to the analysis and presentation of scientific data in support of multi-million dollar partnerships with pharmaceutical companies and academic collaborators

<div class="container">
  <div class="d-flex justify-content-between mx-auto" style="width: 100%;">
    <div class="text-start">
      <p><strong>McDonnell Genome Institute, Washington University Department of Medicine</strong><br>Position: Staff Scientist</p>
    </div>
    <div class="text-end">
      <p>St. Louis, MO<br>Nov 2013 - Aug 2022</p>
    </div>
  </div>
</div>

###### 2020 - 2022 Highlights
- Lead author and analyst for a study of non-cirrotic hepatocellular carcinoma (WGS/WES/Custom Capture) PMID: [35568002](https://pubmed.ncbi.nlm.nih.gov/35568002/)
- Lead author and analyst for a treatment naive/relapse study of small cell lung cancer (WGS/WES/RNAseq) (Pending Publication)
- Lead author and analyst of a protocol for scTCRseq and scRNAseq for canine samples PMID: [38649520](https://pubmed.ncbi.nlm.nih.gov/38649520/)
- Implemented modifications to existing H. sapiens workflows to be compatible with multiple species PMID: [33556121](https://pubmed.ncbi.nlm.nih.gov/33556121/)
- Built a custom workflow to detect viral integration sites from dual indexed viral reads PMID: [34019588](https://pubmed.ncbi.nlm.nih.gov/34019588/)
- Implemented machine learning algorithm to classify and refine somatic variant calls using extreme gradient boosting models
- Optimized analysis scripts for the [RegTools](https://regtools.readthedocs.io/en/latest/) software package to be faster and less memory intensive.

###### 2018 - 2019 Highlights
- Created and instituted Common Workflow Language (CWL) bioinformatic pipelines: [bioinformatic-workflows](https://github.com/griffithlab/bioinformatic-workflows); [analysis-workflows](https://github.com/genome/analysis-workflows)
- Created and maintained [docker images](https://hub.docker.com/u/zlskidmore) for bioinformatic analysis 
- Trained staff scientists in various bioinformatic and programming skill sets.
- Created static websites and subsequent material for genomic visualization, precision medicine, and RNAseq. 
- Advised and/or performed analysis on TCR clonotype repertoire, HLA prediction, neoantigen prediction, ctDNA, WGS, WES, RNAseq PMIDs: [31390566](https://pubmed.ncbi.nlm.nih.gov/31390566/), [31618044](https://pubmed.ncbi.nlm.nih.gov/31618044/), [31467059](https://pubmed.ncbi.nlm.nih.gov/31467059/), [31685621](https://pubmed.ncbi.nlm.nih.gov/31685621/), [31902496](https://pubmed.ncbi.nlm.nih.gov/31902496/), [32665297](https://pubmed.ncbi.nlm.nih.gov/32665297/), [34036230](https://pubmed.ncbi.nlm.nih.gov/34036230/), [33556121](https://pubmed.ncbi.nlm.nih.gov/33556121/), [33547198](https://pubmed.ncbi.nlm.nih.gov/33547198/)
- Prototyped shiny application for interrogation of saturation mutagenesis assays [saturation-mutagenesis-viz](https://github.com/griffithlab/saturation-mutagenesis-viz)

###### 2013 - 2017 Highlights
- Led development and publication of Genomic Visualization in R (GenVisR) library on bioconductor DOI: [10.18129/B9.bioc.GenVisR](https://doi.org/doi:10.18129/B9.bioc.GenVisR) PMIDs: [27288499](https://pubmed.ncbi.nlm.nih.gov/27288499/), [34506690](https://pubmed.ncbi.nlm.nih.gov/34506690/)
- Consulted on design and implementation of Clinical Interpretations of Variants in Cancer [(CIViC)](https://civicdb.org) database; Co-organized first Biohackathon for CIViC at the Netherlands Cancer Institute; Submitted and curated database entries PMID: [28138153](https://pubmed.ncbi.nlm.nih.gov/28138153/)
- Mentored and assisted graduate and undergraduate students in analyzing projects and writing manuscripts.
- Assisted in the maintenance and improvement of institute project management and analysis software (perl) PMID: [26645048](https://pubmed.ncbi.nlm.nih.gov/26645048/)
- Conducted statistical and bioinformatic analysis for next generation sequencing data (Benchmarking Alignment, Clonotype Expansion, Cancer Genomic Alterations, etc.) PMID: [26531824](https://pubmed.ncbi.nlm.nih.gov/26531824/), [26563128](https://pubmed.ncbi.nlm.nih.gov/26563128/), [27058228](https://pubmed.ncbi.nlm.nih.gov/27058228/), [27029710](https://pubmed.ncbi.nlm.nih.gov/27029710/), [27181063](https://pubmed.ncbi.nlm.nih.gov/27181063/), [27681435](https://pubmed.ncbi.nlm.nih.gov/27681435/)
- Performed next generation sequencing analysis and presented results for cancer genomic tumor board n-of-1 cases

### Workshops, Instruction & Community Leadership:
<div class="container">
  <div class="d-flex justify-content-between mx-auto" style="width: 100%;">
    <div class="text-start">
      <p>Teaching assistant/content author for <a href="https://cancer-research-institute.github.io/CRI_bioinformatics_workshop/">Cancer Research Institute bioinformatics workshop</a></p>
    </div>
    <div class="text-end">
      <p>2024, 2025</p>
    </div>
  </div>
</div>

<div class="container">
  <div class="d-flex justify-content-between mx-auto" style="width: 100%;">
    <div class="text-start">
      <p>Instructor/content author for <a href="https://genviz.org">genomic visualizations workshop</a><br>- <a href="https://www.physalia-courses.org/">Physalia-courses</a>, Freie Universität Berlin, Germany<br>- <a href="http://evomics.org/">Evomics, Workshop on Genomics</a>, Český Krumlov, Czech Republic</p>
    </div>
    <div class="text-end">
      <p><br>2017, 2019<br>2019, 2020</p>
    </div>
  </div>
</div>

<div class="container">
  <div class="d-flex justify-content-between mx-auto" style="width: 100%;">
    <div class="text-start">
      <p>Teaching assistant/contributing author for <a href="https://pmbio.org">Precision Medicine workshop</a><br>- PR Informatics, Glasgow, Scotland</p>
    </div>
    <div class="text-end">
      <p><br>2018</p>
    </div>
  </div>
</div>

<div class="container">
  <div class="d-flex justify-content-between mx-auto" style="width: 100%;">
    <div class="text-start">
      <p>Contributing author for <a href="https://rnabio.org">RNAseq workshop</a></p>
    </div>
    <div class="text-end">
      <p></p>
    </div>
  </div>
</div>

### Selection of Scientific Expertise

###### cfDNA Fragmentomics & Liquid Biopsy Data Analysis

- shallow-WGS Fragmentomic Profiling
- shallow-WGS Aneuploidy Detection
- Nucleosome Footprinting
- Early Cancer Detection and Modeling

###### Next-Generation Sequencing Analysis

- Pre/Post Alignment QC (e.g. FastQC, Picard, samtools)
- Read Alignment (Burrows-Wheeler & K-mer Based)
- Somatic/Germline Variant Calling
- Variant Annotation and Filtering (e.g. VEP)
- CNA/SV Detection
- LOH Detection
- Illumina-based WES/WGS/RNA-seq workflows
- UMI-Aware Processing (e.g. fgbio)

###### Cancer Genomics & Tumor Biology Bioinformatics

- Liquid and Solid Tumors
- Tumor Heterogeneity and Clonal Evolution
- Pathway Analysis
- Genomic Visualization and Interpretation
- Tumor Purity Estimation
- Data Mining and Large Cohort Processing

###### RNAseq Analysis

- Differential Expression
- GSEA
- Transcript Quantification
- Alternative Splicing
- scRNA-seq

###### Statistical Analysis

- Hypothesis Testing (Parametric/Non-Parametric)
- Survival Analysis (Kaplan-Meier)
- Batch Effect Detection/Correction
- Power Analysis
- Distribution-based Modeling (e.g., binomial, negative binomial, poisson)

###### Cancer Immunotherapy Data Analysis

- Checkpoint Inhibitor Biomarker Profiling
- TCR Sequencing and Clonotype Repertoire Diversity
- HLA-Typing (e.g. OptiType)
- Neoantigen prediction (e.g. pVACtools)
- Immune Cell Deconvolution (e.g. CIBERSORT)
- TIL quantification from RNAseq

###### Machine Learning and Modeling in Genomics

- Random Forest
- Extreme Gradient Boosting
- Generalized Linear Models
- Cross-Validation
- Model Evaluation (ROC, AUC, Precision-Recall)

###### Experimental Design

- Cohort Stratification
- Univariate and Multivariate Analysis
- Longitudinal Study Frameworks
- Prospective/Retrospective Study Design

### Selection of Technical Skills

###### Advanced

- R (inc. ggplot2, data.table, S4)
- UNIX
- GNU Utils (incl. awk, sed, vim)
- Docker (incl. Image building)
- Git/GitHub
- High Performance Computing (e.g. LSF, Slurm)
- Amazon Web Services (incl. EC2, S3)
- Confluence/JIRA

###### Proficient

- Common Workflow Language and Pipeline Implementation
- Perl
- Jekyll/Static Site Building
- HTML/Web Development

###### Familiar

- SQL
- Python
- CSS/SASS
- D3.js





